server/src/main/java/org/elasticsearch/snapshots/package-info.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * <p>This package exposes the Elasticsearch Snapshot functionality.</p>
+ *
+ * <h1>Preliminaries</h1>
+ *
+ * <p>There are two communication channels between all nodes and master in the snapshot functionality:</p>
+ * <ul>
+ * <li>The master updates the cluster state by adding, removing or altering the contents of its custom entry
+ * {@link org.elasticsearch.cluster.SnapshotsInProgress}. All nodes consume the state of the {@code SnapshotsInProgress} and will start or
+ * abort relevant shard snapshot tasks accordingly.</li>
+ * <li>Nodes that are executing shard snapshot tasks report either success or failure of their snapshot task by submitting a
+ * {@link org.elasticsearch.snapshots.SnapshotShardsService.UpdateIndexShardSnapshotStatusRequest} to the master node that will
+ * update the snapshot's entry in the cluster-state accordingly.</li>
+ * </ul>
+ *
+ * <h1>Snapshot Creation</h1>
+ * <p>Snapshots are created by the following sequence of events:</p>
+ * <ol>
+ * <li>An invocation of {@link org.elasticsearch.snapshots.SnapshotsService#createSnapshot} enqueues a cluster state update to create
+ * a {@link org.elasticsearch.cluster.SnapshotsInProgress.Entry} in the cluster-state's {@code SnapshotsInProgress}.
+ * This initial snapshot entry has its state set to {@code INIT} and an empty map set for the state of the individual shard's snapshots.
+ * </li>
+ *
+ * <li>After the snapshot's entry with state {@code INIT} is in the cluster-state,
+ * {@link org.elasticsearch.snapshots.SnapshotsService} determines the primary shards' assignments for all indices that are being
```

Comment:
This raise the question of what happens if a primary shard moves in between the creation of the `SnapshotsInProgress` entry and the time when it is populated by the `SnapshotsService`. It'd be good to hear more on that.

Reply:
I added a sentence on that in a699b72, hope it helps. Relocation  between `INIT` and `STARTED` isn't really an issue since the assignment of nodes and moving to `STARTED` happens in the same `ClusterStateUpdateTask` (Maybe make that more explicit in the docs? I opted not to since it seemed too detailed here but maybe not?).

Reply:
Thanks, the addition is helpful. Is it right that we do not start to _gracefully_ relocate a primary while it is being snapshotted (i.e. we have `SnapshotInProgressAllocationDecider`), so that if the primary moves later then it's a genuine failure?

What do we do with a primary that is already being relocated when the snapshot starts? Do we use the relocation target (which is initialising, so goes into the `WAITING` state)?

I'm still getting to grips with all of this, so perhaps I'm going into too much detail.

Reply:
>  Is it right that we do not start to _gracefully_ relocate a primary while it is being snapshotted (i.e. we have `SnapshotInProgressAllocationDecider`), so that if the primary moves later then it's a genuine failure?

Yes, that's exactly right :) I felt like it was more confusing than helpful to add the "not graceful" qualifier to the text, but maybe I'm wrong?

Yea exactly, we have this in `SnapshotsService`:

What do we do with a primary that is already being relocated when the snapshot starts? Do we use the relocation target (which is initialising, so goes into the `WAITING` state)?

```java
 } else if (primary.relocating() || primary.initializing()) {
   builder.put(shardId, new SnapshotsInProgress.ShardSnapshotStatus(primary.currentNodeId(), State.WAITING));
```